nengo_loihi/emulator/emulator.py

```diff
@@ -40,137 +39,364 @@ class Emulator(object):
-        else:
-            warnings.warn(msg)
+    def close(self):
+        self.closed = True
 
-    def build(self, model, seed=None):  # noqa: C901
-        """Set up NumPy arrays to emulate chip memory and I/O."""
-        model.validate()
+        # remove references to states to free memory (except probes)
+        self.group_info = None
+        self.inputs = None
+        self.compartments = None
+        self.synapses = None
+        self.axons = None
 
-        if seed is None:
-            seed = np.random.randint(2**31 - 1)
+    def chip2host(self, probes_receivers=None):
+        if probes_receivers is None:
+            probes_receivers = {}
 
-        logger.debug("Emulator seed: %d", seed)
-        self.seed = seed
-        self.rng = np.random.RandomState(seed)
+        increment = 0
+        for probe, receiver in probes_receivers.items():
+            inc = self.probes.send(probe, self._chip2host_sent_steps, receiver)
+            increment = inc if increment == 0 else increment
+            assert inc == 0 or increment == inc
 
-        self.t = 0
+        self._chip2host_sent_steps += increment
 
-        self.model = model
-        self.inputs = list(self.model.inputs)
-        self.groups = list(self.model.groups)
-        self.probe_outputs = {}
-        for obj in self.inputs + self.groups:
-            for probe in obj.probes:
-                self.probe_outputs[probe] = []
-
-        self.n_cx = sum(group.n for group in self.groups)
-        self.group_cxs = {}
-        i0, i1 = 0, 0
-        for group in self.groups:
-            i1 = i0 + group.n
-            self.group_cxs[group] = slice(i0, i1)
-            i0 = i1
+    def host2chip(self, spikes, errors):
+        for spike_input, t, spike_idxs in spikes:
+            spike_input.add_spikes(t, spike_idxs)
+
+        self.synapses.update_pes_errors(errors)
+
+    def run_steps(self, steps):
+        """Simulate for the given number of ``dt`` steps.
+
+        Parameters
+        ----------
+        steps : int
+            Number of steps to run the simulation for.
+        """
+        for _ in range(steps):
+            self.step()
+
+    def step(self):
+        """Advance the simulation by 1 step (``dt`` seconds)."""
+        self.t += 1
+        self.compartments.advance_input()
+        self.synapses.inject_current(
+            self.t, self.inputs, self.axons, self.compartments.spiked)
+        # self.compartments.update_input(self.synapses)
+        self.synapses.update_input(self.compartments.input)
+        self.synapses.update_traces(self.t, self.rng)
+        self.synapses.update_weights(self.t, self.rng)
+        self.compartments.update(self.rng)
+        self.probes.update(self.t, self.compartments)
 
-        # --- allocate group memory
-        group_dtype = self.groups[0].vth.dtype
-        assert group_dtype in (np.float32, np.int32)
+    def get_probe_output(self, probe):
+        return self.probes[probe]
+
+
+class GroupInfo(object):
```

Comment:
Could you add docstrings for all the new state classes (GroupInfo, IterableState, CompartmentState, NoiseState, AxonState, SynapseState, ProbeState)?

Reply:
I've added some docstrings. I didn't document attributes for `CompartmentGroup` and `NoiseGroup`, because there are a lot and they map pretty directly from the corresponding values in `Compartments`.